app/controllers/admin/posts_controller.rb

```diff
@@ -93,7 +88,8 @@ def load_element
   end
 
   def post_include
-    @posts = Post.includes :user, :topic
+    @q = Post.ransack params[:q]
+    @posts = @q.result.includes :user, :topic
```

Comment:
tại sao dùng includes e nhỉ
sao k dùng eager load vs preload

Reply:
em tưởng là mặc định của includes sẽ hoạt động theo cơ chế select in của thằng preload ạ :>

Reply:
tìm hiểu lại 3 cái đó và trả lời vào đây nhé

Reply:
Có 3 cách để giải quyết vấn đề n + 1 query : includes, preload, eagerload 
 Giả sử ta có ví dụ 1 user có nhiều bài post: 
 Cách 1 : Preload
 -  Sử dụng cơ chế của select_in() dùng 2 câu lệnh truy vấn :
    1. Truy vấn đầu tiên để lấy hết toàn bộ user.  
    2. Truy vấn thứ hai để load các post tương ứng với những users đó.
Cách 2: Eagerload
- Sử dụng cơ chế của joins dùng 1 câu lệnh truy vấn:
  1. Thực hiện việc join table user và table post .
  2. Truy vấn lấy ra các bản ghi .
Cách 3: Includes
 1.Mặc định của includes sẽ hoạt động theo cơ chế của select_in() -> User.includes(:posts).
 2.Khi ta cần tìm kiếm theo thuộc tính của post nhưng select_in() lại không thể load trước được dữ liệu từ bảng post thì ta phải sự dụng cơ chế của joins bằng cách : User.includes(:posts).references(:posts).
- Đối với quá nhiều bản ghi ( Ví dụ: 1 triệu bản )
 + Nếu ta sử dụng được select_in() thì sử dụng select_in().
 + Nếu ta không thể sử dụng select_in() thì ta nên sử dụng N+1 query thay vì ta dùng joins vì joins sẽ làm giảm performance